### ISSUE TITLE
The mounted volume /opt/graphite contains the graphite application and libraries, not just config and logs

### ISSUE THREAD

**Issue by dbeckham** [hopsoft/docker-graphite-statsd#43]
What is the reason for making `/opt/graphite` a docker volume instead of just the storage and log directories by themselves? The graphite binaries are installed in `/opt/graphite`, not just the config files and logs. If you mount that volume on the host, or in a docker volume container, what happens when you update the image to graphite 0.9.14?  How will the scripts in `/opt/graphite/bin` and `/opt/graphite/lib` be updated?

**Comment by jareware** [comment]
Came here to open a very related issue, so I'll just add it here: because `/opt/graphite` is one single volume, it's very hard to specify that by hand. So let's say you'd want to keep the storage area somewhere where it's easy to back up, you could e.g. `docker run ... -v /home/jareware/graphite:/opt/graphite ...`. But you can't, because then the container won't start because that directory will be empty by default.

**Comment by dbeckham** [comment]
Right.  Docker volumes just don't work the way that the Dockerfile seems to use them. The problem is even more exacerbated by the fact that a traditional docker storage volume is almost always, not the right place to store graphite metrics. I could see volumes for configurations and logs, but not for application data. 

What I've ended up doing is just forking this setup and removing the `/opt/graphite` volume so that I can manage the configuration and whisper file data in the right way.

**Comment by jareware** [comment]
> traditional docker storage volume is almost always, not the right place to store graphite metrics

OK now I'm curious; what would be the correct way?

**Comment by jareware** [comment]
Also @dbeckham, if your fork is public I might be interested in checking it out! :)

**Comment by dbeckham** [comment]
> OK now I'm curious; what would be the correct way?

@jareware in my mind I was thinking of the traditional aufs or overlay filesystem use, but docker volumes do bypass those slower filesystems. The issue mostly remains that /opt/graphite (and /opt/statsd) contains a mix of both graphite application binaries and libraries, and configuration files which isn't appropriate to use with a `VOLUME ...` command in a Docker file.

>Also @dbeckham, if your fork is public I might be interested in checking it out! :)

Our busiest time of the year is during Black Friday, so I've been focused on our site performance, but I'm working on pushing my changes up to a branch in my fork. Most of what I'm doing is compatible with the master branch here, but there are a few things that may not be. 

The biggest changes are:
* Updated to Graphite 0.9.15 
* Added ability to disable unneeded or unused services through environment variables (Statsd isn't really part of graphite, even if it's ubiquitous, and scaling may require it to run from a different location or even different docker container.)
* Removed `/opt/graphite` and `/opt/statsd` volumes from dockerfile. It's expected that nonstandard configs are handled by `-v/--volume` options in your `docker run` command.
* Fixed several bugs: service control; incorrect carbon storage schemas; exposed missing ports; other fixes
* Added support to local_settings.py for reading in an external config file so that custom settings like authentication secrets, email config, timezone, memcache servers, etc., can be configured without having to do a special mount to `/opt/graphite/webapp/graphite/local_settings.py` in the `docker run` command.

I'm hoping I'll be able to push my changes up in the next few days.

**Comment by hopsoft** [comment]
:+1: I'm looking forward to seeing your changes.

**Comment by jareware** [comment]
@dbeckham oooh, looking forward to it!

As to the configs, there's also #44, but a simple overrides file would be nice as well.

**Comment by dbeckham** [comment]
@jareware and @hopsoft I just pushed my fork's changes (which were put into production earlier today) to a 0.9.x branch: https://github.com/dbeckham/docker-graphite-statsd/tree/0.9.x  

There are a few outstanding issues related to logging for Docker logging driver support, but they need some thought before implementing.

Changelog:

 * New: Upgraded to Graphite 0.9.15
 * New: Upgraded to latest phews/baseimage Docker image (v0.9.17)
 * New: Upgraded Nginx to latest mainline release (1.8)
 * New: Add support for disabling individual services through environment variables
 * New: Added support to local_settings.py for including an external file for custom configuration settings.
 * Change: replaced carbon `--debug` option with `--nodaemon`
 * Fix: Add missing exposed ports for carbon, carbon-aggregator and statsd
 * Fix: Correct carbon retention periods to match carbon.conf settings
 * Fix: Services weren't being restarted correctly by runit
 * Fix: Remove Docker volumes created for application binaries and libraries
 * Fix: Add startup script to ensure that `/var/log/*` directories exist for services that do not create them
 * Fix: Changed Docker WORKDIR to / instead of the last /usr/src directory a package was installed from

**Comment by jareware** [comment]
Awesome work!! :bow:

@hopsoft do you plan on pulling these changes here?

**Comment by dbeckham** [comment]
@jareware Some changes may be incompatible or too overreaching. I didn't specifically try to make things not be mergeable, but I was originally making changes for our own needs and long term use. After you asked if it would be publicably available I reversed a few changes and made them configurable, like statsd being installed and any Dockerfile volumes existing.

If any of these changes are worth merging, I don't mind repackaging them into a branch for easier pulls.

**Comment by hopsoft** [comment]
I really like the changes & would like to pull them into this project. @dbeckham Can you split the change set up into a series of smaller more isolated concerns/pull requests? That should make things easier to review & get merged quickly. Thanks for your work to improve things.

**Comment by dbeckham** [comment]
Sure, I'd be glad to. It may take me a few days to work through them, but I'll start putting them into logical patches.

**Comment by arnaudbey** [comment]
@hopsoft @dbeckham 
Any news on this ? 
Or any tips to get some data persistence (i mean possibility of store datas outside the container and retrieve it after a docker stop/start) ?

I've tried both this repo and dbeckam's one (the 0.9.x branch), and can't make it work. 

Thanks for your time/work ! :+1: 

Thank you both for your time

**Comment by rmelick** [comment]
I just ran into issues when restarting my container with storing data in an external docker volume.  The issue I saw was that the carbon-cache processes would not come up after the container was restarted.  

After starting up with this command
```
sudo docker run -d\
 --name graphite\
 --restart=always\
 -p 80:80\
 -p 2003-2004:2003-2004\
 -p 2023-2024:2023-2024\
 -p 8125:8125/udp\
 -p 8126:8126\
 -v /opt/graphite/storage:/opt/graphite/storage\
 hopsoft/graphite-statsd
```
I restarted using `sudo docker restart graphite`.   Then, I saw errors like this (executed from the host, not inside the docker container).

```
ubuntu@ip-10-0-198-200:/opt/graphite/conf$ telnet localhost 2004
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.
```
After I removed the carbon-aggregator-a.pid and carbon-cache-a.pid inside of /opt/graphite/storage, I was able to connect successfully.

Any thoughts on the best way to mount a storage volume but not the pid files?  Or would it make sense to have logic to delete the pids on startup?

**Comment by hopsoft** [comment]
Thanks for the report on issues with pid files on restart. I'll look into deleting these on container start.
